### PR TITLE
feat: add `scopeKey` config option to hooks

### PR DESCRIPTION
**Pull request by jxom**
## Description

This PR adds a new configuration option to our "query" hooks, `scopeKey`, which scope the hook's cache to a given context. This key will be useful for situations like #1127 where consumers may want to have hooks which are more "pure" (ie. the result of a hook shouldn't affect other instances of it). The `scopeKey` essentially gives the consumer access to add on an arbitrary key to the query key.

### CODE REVIEW

File: packages/react/src/types/index.ts [hunk 1]
```diff
@@ -77,7 +80,10 @@ export type InfiniteQueryConfig<Data, Error> = Pick<
   | 'onError'
   | 'onSettled'
   | 'onSuccess'
->
+> & {
+  /** Scope the cache to a given context. */
+  contextKey?: string
```
Comment (tmm) [inline packages/react/src/types/index.ts:85]: Wonder if it's better to call this something else so it doesn't get confused with React Context or React Query's `context` option if we choose to add something similar in the future? Maybe `scopeKey` or `cacheKey`?

Comment (jxom) [reply packages/react/src/types/index.ts:85]: Yeah, maybe `scopeKey` is a better fit. I have a feeling people may take `cacheKey` too literally (ie. think it is the _actual_ query key).

File: packages/react/src/hooks/accounts/useBalance.test.ts [hunk 2]
```diff
@@ -82,6 +82,37 @@ describe('useBalance', () => {
...
+        const { result, waitFor } = renderHook(() => {
+          return {
+            balance: useBalance({
+              addressOrName: '0xA0Cf798816D4b9b9866b5330EEa46a18382f251e',
+            }),
+            balanceWithoutContextKey: useBalance({
+              addressOrName: '0xA0Cf798816D4b9b9866b5330EEa46a18382f251e',
+              enabled: false,
+            }),
+            balanceWithContextKey: useBalance({
```
Comment (tmm) [inline packages/react/src/hooks/accounts/useBalance.test.ts:95]:
```suggestion
            balanceWithScopeKey: useBalance({
```

File: packages/react/src/hooks/contracts/useContractRead.ts [hunk 3]
```diff
@@ -64,6 +69,8 @@ function queryFn<
   }: QueryFunctionArgs<typeof queryKey>) => {
     if (!abi) throw new Error('abi is required')
     if (!address) throw new Error('address is required')
+    if (!args) throw new Error('args is required')
```
Comment (tmm) [inline packages/react/src/hooks/contracts/useContractRead.ts:72]: `args` isn't always required (e.g. wagmigotchi `getHunger`)

**Review by tmm — CHANGES_REQUESTED** [review]

**Review by tmm — APPROVED** [review]

**Pull request merged**